core/drivers/stm32_rng.c

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2018, STMicroelectronics - All Rights Reserved
+ */
+
+#include <assert.h>
+#include <drivers/stm32_rng.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <kernel/delay.h>
+#include <kernel/generic_boot.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <stdbool.h>
+#include <stm32_util.h>
+#include <string.h>
+
+#ifdef CFG_DT
+#include <libfdt.h>
+#endif
+
+#define DT_RNG_COMPAT		"st,stm32-rng"
+#define RNG_CR			0x00U
+#define RNG_SR			0x04U
+#define RNG_DR			0x08U
+
+#define RNG_CR_RNGEN		BIT(2)
+#define RNG_CR_IE		BIT(3)
+#define RNG_CR_CED		BIT(5)
+
+#define RNG_SR_DRDY		BIT(0)
+#define RNG_SR_CECS		BIT(1)
+#define RNG_SR_SECS		BIT(2)
+#define RNG_SR_CEIS		BIT(5)
+#define RNG_SR_SEIS		BIT(6)
+
+#define RNG_TIMEOUT_US		1000
+
+struct stm32_rng_instance {
+	struct io_pa_va base;
+	unsigned long clock;
+};
+
+static struct stm32_rng_instance *stm32_rng;
+
+int stm32_rng_read_raw(uintptr_t rng_base, uint8_t *out, size_t size)
+{
+	uint8_t *buf = out;
+	size_t len = size;
+	uint64_t timeout_ref = 0;
+	int count = 0;
+
+	if (!(read32(rng_base + RNG_CR) & RNG_CR_RNGEN))
+		write32(RNG_CR_RNGEN | RNG_CR_CED, rng_base + RNG_CR);
```

Comment:
What does this do? Enable RNG generation for a few words or "forever"?

Reply:
I'll add a comment. Indeed this enables forever. I will add a disable sequence once data are read.